### PR TITLE
Add get patient by id from FHIR module endpoint

### PR DESCRIPTION
**Pull request by cecilyshentw**
Adding an endpoint to retrieve Patient by id from FHIR
This is a preparation for card #41.

### CODE REVIEW

File: test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+            ushort?  patientYob = 1981;
+            var patientEnquiry =
+                new PatientEnquiry(
+                    "id", verifiedIdentifiers: null, unverifiedIdentifiers: null,
+                    patientName, patientGender, patientYob);
+            var request = new DiscoveryRequest(patientEnquiry,"requestId", "transactionId", DateTime.Now);
+            var repo = new OpenMrsPatientMatchingRepository(patientDal.Object);
+           
+            var result = repo.Where(request);
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs:59]: Codacy found an issue: [Remove the unused local variable 'result'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 27
           

Complexity increasing per file
==============================
- test/In.ProjectEKA.HipServiceTest/Common/TestServer/TestAuthHandler.cs  1
- src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs  1
- src/In.ProjectEKA.HipLibrary/Validation/AtLeastOneHasValueValidation.cs  5
- test/In.ProjectEKA.HipServiceTest/Discovery/PatientControllerTest.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs  1
- test/In.ProjectEKA.HipServiceTest/Common/TestServer/TestStartup.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs  1
- src/In.ProjectEKA.HipService/Discovery/Matcher/ExactNameMatcher.cs  1
- test/In.ProjectEKA.HipServiceTest/Common/TestServer/AuthenticatedTestRequestMiddleware.cs  1
- src/In.ProjectEKA.HipService/OpenMrs/OpenMrsConfiguration.cs  1
- src/In.ProjectEKA.HipService/OpenMrs/OpenMrsPatientMatchingRepository.cs  7
- test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs  1
- test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/Builder/DiscoveryRequestPayloadBuilder.cs  6
- test/In.ProjectEKA.HipServiceTest/OpenMrs/OpenMrsClientTest.cs  3
- src/In.ProjectEKA.HipService/OpenMrs/Mappings/PatientExtensions.cs  4
- src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs  6
- src/In.ProjectEKA.HipService/Discovery/Ranker/GenderRanker.cs  1
- src/In.ProjectEKA.HipService/Discovery/Ranker/ExactNameRanker.cs  1
- src/In.ProjectEKA.HipService/Discovery/Matcher/DemographicMatcher.cs  1
         

Clones added
============
- src/In.ProjectEKA.HipService/Startup.cs  1
- test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs  2
- test/In.ProjectEKA.HipServiceTest/Common/TestServer/TestStartup.cs  1
- test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs  3
- test/In.ProjectEKA.HipServiceTest/OpenMrs/OpenMrsClientTest.cs  3
         

Clones removed
==============
+ test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330&bid=19443097)

File: README.md [hunk 2]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+To make them work, you first need to get a authentication token from the oidc-server-mock by providing the configured clientid, client secret, grant type and scope (see these settings in the client configuration file: /gatewayStubConfigurations/oicdConfigs/clients-config.json). Using this authentication token as an Oath 2.0 token, you can send a request to the HIP service that will be accepted by the authentication.
+Please note in the docker compose file the ports that these containers need to use to be able to run. 
+You can use the described flow with any tooling, however, we included a sample Postman collection (/gatewayStubConfigurations/postmanCollections/HIPSamplePostmanCollections.json) that you can use for sending the requests, or for reference.
+Please follow the following steps with the Postman collection:
+1. make sure that the docker stack is running - in case needed, (re)build docker stack (docker-compose -f docker-compose.yml up --build -d)
+2. open postman and import the collection
+3. if you don't have, create an environment in postman (https://learning.postman.com/docs/sending-requests/managing-environments/)
```
Comment (ganesanarun) [inline README.md:105]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md [hunk 3]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+4. run the Post Token request from the imported collection
```
Comment (ganesanarun) [inline README.md:106]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md [hunk 4]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+'oidc-server-mock' is used to simulate the OpenID Connect authentication provided by the Gateway service.
+To make them work, you first need to get a authentication token from the oidc-server-mock by providing the configured clientid, client secret, grant type and scope (see these settings in the client configuration file: /gatewayStubConfigurations/oicdConfigs/clients-config.json). Using this authentication token as an Oath 2.0 token, you can send a request to the HIP service that will be accepted by the authentication.
+Please note in the docker compose file the ports that these containers need to use to be able to run. 
+You can use the described flow with any tooling, however, we included a sample Postman collection (/gatewayStubConfigurations/postmanCollections/HIPSamplePostmanCollections.json) that you can use for sending the requests, or for reference.
+Please follow the following steps with the Postman collection:
+1. make sure that the docker stack is running - in case needed, (re)build docker stack (docker-compose -f docker-compose.yml up --build -d)
```
Comment (ganesanarun) [inline README.md:103]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+            openmrsClientMock.Verify(client => client.GetAsync(expectedPath), Times.Once);
+        }
+
+        [Fact(Skip = "Requires AWS access")]
+        [Trait("Category", "Infrastructure")]
+        public async System.Threading.Tasks.Task ShouldGetPatientDataRealCallAsync()
+        {
+            //Given
+            // Disable SSL verification in test only
+            var handler = new HttpClientHandler()
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs:141]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    public class PatientExtensionsTest
+    {
+        const string patientName = "Patient name";
+
+        [Fact]
+        private void ToHipPatient_GivenOpenMrsPatientWithMultipleNames_UsesSearchedName()
+        {
+            var openMrsPatient = new OpenMrsPatient() {
+                Name = new List<OpenMrsPatientName>{  new OpenMrsPatientName() { Text = $"OpenMRS {patientName}" }, new OpenMrsPatientName() { Text = "a second name" } },
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs:22]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs [hunk 7]
```diff
@@ -1,100 +1,168 @@
...
     public class FilterTest
     {
+        private class Linda
+        {
+            public const string Name = "Linda";
+            public const Gender Sex = Gender.F;
+            public const string PhoneNumber = "99999999999";
+            public const ushort YearOfBirth = 1972;
+        }
+        private class John
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs:23]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: src/In.ProjectEKA.HipService/Discovery/Ranker/ExactNameRanker.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    using HipLibrary.Patient.Model;
+    using Matcher;
+    using static RankBuilder;
+    using static MetaBuilder;
+
+    public class ExactNameRanker : IRanker<Patient>
+    {
+        public PatientWithRank<Patient> Rank(Patient patient, string name)
+        {
+            return (ExactNameMatcher.IsMatch(patient.Name, name)) switch
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/Discovery/Ranker/ExactNameRanker.cs:12]: Codacy found an issue: [Replace this 'switch' expression with a ternary conditional operator to increase readability.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    using System.Linq;
+    using System.Threading.Tasks;
+    using OpenMrsPatient = Hl7.Fhir.Model.Patient;
+    using OpenMrsPatientName = Hl7.Fhir.Model.HumanName;
+    using OpenMrsGender = Hl7.Fhir.Model.AdministrativeGender;
+    using In.ProjectEKA.HipService.OpenMrs;
+
+    public class OpenMrsPatientMatchingRepositoryTest
+    {
+        private Mock<IPatientDal> patientDal = new Mock<IPatientDal>();
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs:18]: Codacy found an issue: [Make 'patientDal' 'readonly'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            hipGender.ToOpenMrsGender().Should().Be(openMrsGender);
+        }
+
+        [Theory]
+        [InlineData("1")]
+        [InlineData("12345678-1234-1234-1234-123456789abc")]
+        [InlineData(null)]
+        private void ToHipPatient_GivenOpenMrsPatient_IdIsMappedToPatientId(string patientId)
+        {
+            var openMrsPatient = new OpenMrsPatient()
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs:91]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md [hunk 11]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+2. open postman and import the collection
+3. if you don't have, create an environment in postman (https://learning.postman.com/docs/sending-requests/managing-environments/)
+4. run the Post Token request from the imported collection
+5. run the Happy Case request from the imported collection -> you should get a 202 response
+6. if you go go to the logs for the gateway stub container, you will see the response posted to gateway
```
Comment (ganesanarun) [inline README.md:108]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md (hunk 2 above)
Comment (ganesanarun) [inline README.md:105]: Codacy found an issue: [[no-literal-urls] Don’t use literal URLs without angle brackets](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md [hunk 12]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+'oidc-server-mock' is used to simulate the OpenID Connect authentication provided by the Gateway service.
+To make them work, you first need to get a authentication token from the oidc-server-mock by providing the configured clientid, client secret, grant type and scope (see these settings in the client configuration file: /gatewayStubConfigurations/oicdConfigs/clients-config.json). Using this authentication token as an Oath 2.0 token, you can send a request to the HIP service that will be accepted by the authentication.
+Please note in the docker compose file the ports that these containers need to use to be able to run. 
+You can use the described flow with any tooling, however, we included a sample Postman collection (/gatewayStubConfigurations/postmanCollections/HIPSamplePostmanCollections.json) that you can use for sending the requests, or for reference.
+Please follow the following steps with the Postman collection:
+1. make sure that the docker stack is running - in case needed, (re)build docker stack (docker-compose -f docker-compose.yml up --build -d)
+2. open postman and import the collection
```
Comment (ganesanarun) [inline README.md:104]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+        public async Task<List<Patient>> LoadPatientsAsync(string name, AdministrativeGender? gender, string yearOfBirth)
+        {
+            var path = DiscoveryPathConstants.OnPatientPath;
+            var query = HttpUtility.ParseQueryString(string.Empty);
+            if (!string.IsNullOrEmpty(name)) {
+                query["name"]=name;
+            }
+            if (gender != null) {
+                query["gender"]=gender.ToString().ToLower();
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/OpenMrs/DiscoveryDataSource.cs:28]: Codacy found an issue: [Define the locale to be used in this string operation.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+                            It.IsAny<string>(),
+                            It.IsAny<OpenMrsGender?>(),
+                            It.IsAny<string>()))
+                        .Returns(
+                            (string name, OpenMrsGender? gender, string yob) => {
+                                var humanName = new OpenMrsPatientName();
+                                humanName.Text = $"OpenMRS name of {name}";
+
+                                return Task.FromResult(
+                                    new List<OpenMrsPatient>() {
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/OpenMrsPatientMatchingRepositoryTest.cs:34]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        [InlineData("1981", (UInt16)1981)]
+        [InlineData("1973-06", (UInt16)1973)]
+        [InlineData("1905-08-23", (UInt16)1905)]
+        [InlineData(null, null)]
+        private void ToHipPatient_GivenOpenMrsPatient_YearOfBirthIsCalculatedFromBirthDate(string sourceBirthDate, ushort? expectedYearOfBirth)
+        {
+            // The hl7 date format is YYYY, YYYY-MM, or YYYY-MM-DD, e.g. 2018, 1973-06, or 1905-08-23. 
+            // https://www.hl7.org/fhir/datatypes.html#date
+
+            var openMrsPatient = new OpenMrsPatient() {
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs:64]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/OpenMrs/OpenMrsClientTest.cs [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    [Collection("Gateway Client Tests")]
+    public class OpenMrsClientTest
+    {
+        [Fact(Skip="it is a real call and needs local setup")]
+        [Trait("Category", "Infrastructure")]
+        public async Task ShouldGetPatientDataRealCallAsync()
+        {
+            //Given
+            // Disable SSL verification in test only
+            var handler = new HttpClientHandler()
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/OpenMrs/OpenMrsClientTest.cs:23]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs [hunk 17]
```diff
@@ -1,100 +1,168 @@
...
-    using Xunit;
+    using HipService.Discovery;
+    using Xunit;
     using static Builder.TestBuilders;
 
+
     [Collection("Patient Filter Tests")]
     public class FilterTest
     {
+        private class Linda
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/FilterTest.cs:16]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+        public async Task<HttpResponseMessage> GetAsync(string openMrsUrl)
+        {
+            try {
+                return await httpClient.GetAsync(Path.Join(configuration.Url, openMrsUrl));
+            }
+            catch (Exception exception)
+            {
+                Log.Error(exception, exception.StackTrace);
+                throw exception;
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/OpenMrs/OpenMrsClient.cs:32]: Codacy found an issue: [Consider using 'throw;' to preserve the stack trace.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+                Name = new List<OpenMrsPatientName>{  new OpenMrsPatientName() { Text = patientName } },
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/Discovery/PatientExtensionsTest.cs:65]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md (hunk 11 above)
Comment (ganesanarun) [inline README.md:108]: Codacy found an issue: [[no-consecutive-blank-lines] Remove 1 line after node](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: src/In.ProjectEKA.HipService/OpenMrs/Mappings/PatientExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,24 @@
+﻿namespace In.ProjectEKA.HipService.OpenMrs.Mappings
+{
+    using HipLibrary.Patient.Model;
+
+    public static class PatientExtensions
+    {
+        public static Patient ToHipPatient(this Hl7.Fhir.Model.Patient openMrsPatient, string patientSearchedName)
+        {
+
+            return new Patient()
```
Comment (ganesanarun) [inline src/In.ProjectEKA.HipService/OpenMrs/Mappings/PatientExtensions.cs:10]: Codacy found an issue: [Remove these redundant parentheses.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: gatewayStubConfigurations/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,8 @@
+FROM andyrbell/mountebank
```
Comment (ganesanarun) [inline gatewayStubConfigurations/Dockerfile:1]: Codacy found an issue: [Always tag the version of an image explicitly](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+                .Setup(x => x.GetAsync(It.IsAny<string>()))
+                .ReturnsAsync(new HttpResponseMessage
+                {
+                    StatusCode = HttpStatusCode.OK,
+                    Content = new StringContent(PatientSample)
+                })
+                .Verifiable();
+
+            //When
+            var patients = await discoveryDataSource.LoadPatientsAsync(name, gender, yearOfBrith);
```
Comment (ganesanarun) [inline test/In.ProjectEKA.HipServiceTest/OpenMrs/DiscoveryDataSourceTest.cs:129]: Codacy found an issue: [Remove the unused local variable 'patients'.](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

File: README.md [hunk 23]
```diff
@@ -90,3 +90,20 @@ Once ran the application, navigate to
...
+3. if you don't have, create an environment in postman (https://learning.postman.com/docs/sending-requests/managing-environments/)
+4. run the Post Token request from the imported collection
+5. run the Happy Case request from the imported collection -> you should get a 202 response
```
Comment (ganesanarun) [inline README.md:107]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 1 space](https://app.codacy.com/gh/ProjectEKA/hip-service/pullRequest?prid=5933330)

**Comment by cecilyshentw** [comment]
Sorry, wrong PR. This should be under Bahmni/hip-service fork

**Pull request closed without merging**